plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchSearchEngine.ts

```diff
@@ -78,90 +87,40 @@ export class ElasticSearchSearchEngine implements SearchEngine {
       aliasPostfix,
       indexPrefix,
       logger,
     );
   }
 
-  private static async constructElasticSearchClient(
-    logger: Logger,
-    config?: Config,
-  ) {
-    if (!config) {
-      throw new Error('No elastic search config found');
-    }
-
-    const clientOptionsConfig = config.getOptionalConfig('clientOptions');
-    const sslConfig = clientOptionsConfig?.getOptionalConfig('ssl');
-
-    if (config.getOptionalString('provider') === 'elastic') {
-      logger.info('Initializing Elastic.co ElasticSearch search engine.');
-      const authConfig = config.getConfig('auth');
-      return new Client({
-        cloud: {
-          id: config.getString('cloudId'),
-        },
-        auth: {
-          username: authConfig.getString('username'),
-          password: authConfig.getString('password'),
-        },
-        ...(sslConfig
-          ? {
-              ssl: {
-                rejectUnauthorized:
-                  sslConfig?.getOptionalBoolean('rejectUnauthorized'),
-              },
-            }
-          : {}),
-      });
-    }
-    if (config.getOptionalString('provider') === 'aws') {
-      logger.info('Initializing AWS ElasticSearch search engine.');
-      const awsCredentials = await awsGetCredentials();
-      const AWSConnection = createAWSConnection(awsCredentials);
-      return new Client({
-        node: config.getString('node'),
-        ...AWSConnection,
-        ...(sslConfig
-          ? {
-              ssl: {
-                rejectUnauthorized:
-                  sslConfig?.getOptionalBoolean('rejectUnauthorized'),
-              },
-            }
-          : {}),
-      });
-    }
-    logger.info('Initializing ElasticSearch search engine.');
-    const authConfig = config.getOptionalConfig('auth');
-    const auth =
-      authConfig &&
-      (authConfig.has('apiKey')
-        ? {
-            apiKey: authConfig.getString('apiKey'),
-          }
-        : {
-            username: authConfig.getString('username'),
-            password: authConfig.getString('password'),
-          });
-    return new Client({
-      node: config.getString('node'),
-      auth,
-      ...(sslConfig
-        ? {
-            ssl: {
-              rejectUnauthorized:
-                sslConfig?.getOptionalBoolean('rejectUnauthorized'),
-            },
-          }
-        : {}),
-    });
+  /**
+   * Re-use the configuration of this search engine in order to construct other
+   * elastic search clients. This is useful if you want to create a client that
+   * talks to the same search cluster as your search engine, but you want to
+   * provide queries and receive results using a completely different format
+   * such as Relay pagination, or faceted search.
+   *
+   * ```javascript
+   * import { Client } from '@elastic/elastic-search';
+   *
+   * let client = engine.newClient(options => new Client(options));
+   * ```
+   */
+  newClient<T>(create: (options: ElasticSearchClientOptions) => T): T {
```

Comment:
This has me thinking this could be useful for other `SearchEngine` implementations (PG or even lunr).  I don't think we need to / should add it to the interface yet.  Just thinking aloud about ensuring the types/API here would be good going into the future, beyond just ES.

Reply:
We could add a `NewClient` interface:

```ts
export interface NewClient<TOptions> {
  newClient<TClient>(create: (options: TOptions) => TClient): TClient;
}
```

And then it can be added to the search engine implementation with:

```ts
export class ElasticSearchSearchEngine implements SearchEngine, NewClient<ElasticSearchClientOptions> {
  /* .... */
}
```

Which would allow it to be incrementally added to each concrete search engine before making it a part of the search engine API proper. I can add that interface locally, or just leave it as a one-off here for now.

Reply:
I like the idea!  But no worries about this for now; something for later if more folks discover it and think it'd be useful in other search engines.